File: backend/app/api/v1/health.py
"""Liveness + readiness endpoints.

  /health : process is up (cheap, used by Docker healthcheck)
  /ready  : downstream services reachable (Postgres + Redis + Chroma, and
            Ollama if it's the configured LLM provider)
"""
from __future__ import annotations

import httpx
from fastapi import APIRouter
from sqlalchemy import text

from app.ai.vectorstore.factory import get_vector_store
from app.api.deps import DbDep
from app.config import settings
from app.core.redis import get_redis_client

router = APIRouter()


@router.get("/health")
async def health() -> dict[str, str]:
    return {
        "status": "ok",
        "app": settings.app_name,
        "env": settings.app_env,
        "llm_provider": settings.llm_provider,
    }


@router.get("/ready")
async def ready(db: DbDep) -> dict:
    components: dict[str, str] = {}

    try:
        await db.execute(text("SELECT 1"))
        components["postgres"] = "ok"
    except Exception as e:
        components["postgres"] = f"error: {e!s}"

    try:
        await get_redis_client().ping()
        components["redis"] = "ok"
    except Exception as e:
        components["redis"] = f"error: {e!s}"

    try:
        await get_vector_store().list_collections()
        components["chroma"] = "ok"
    except Exception as e:
        components["chroma"] = f"error: {e!s}"

    if settings.llm_provider == "ollama":
        try:
            async with httpx.AsyncClient(timeout=3.0) as client:
                r = await client.get(f"{settings.ollama_base_url}/api/tags")
                r.raise_for_status()
            components["ollama"] = "ok"
        except Exception as e:
            components["ollama"] = f"error: {e!s}"

    all_ok = all(v == "ok" for v in components.values())
    return {
        "status": "ready" if all_ok else "degraded",
        "components": components,
    }

File: backend-ts/src/routes/health.ts
/**
 * Liveness + readiness routes.
 *
 *   GET /health   — process is up (no dependencies checked)
 *   GET /ready    — postgres + redis are reachable
 *
 * Mirrors the Python backend so the frontend and any uptime probe keep working.
 */
import type { FastifyInstance } from "fastify";
import { prisma } from "../db.js";
import { config } from "../config.js";
import { redis } from "../lib/cache.js";

export default async function healthRoutes(app: FastifyInstance) {
  app.get("/health", async () => ({
    status: "ok",
    app: config.APP_NAME,
    env: config.APP_ENV,
    llm_provider: config.LLM_PROVIDER,
  }));

  app.get("/ready", async (_req, reply) => {
    const components: Record<string, string> = {};

    try {
      await prisma.$queryRaw`SELECT 1`;
      components.postgres = "ok";
    } catch (e) {
      components.postgres = `error: ${(e as Error).message}`;
    }

    try {
      if (redis.status === "wait" || redis.status === "end") {
        await redis.connect().catch(() => undefined);
      }
      await redis.ping();
      components.redis = "ok";
    } catch (e) {
      components.redis = `error: ${(e as Error).message}`;
    }

    const ready = Object.values(components).every((v) => v === "ok");
    return reply.status(ready ? 200 : 503).send({
      status: ready ? "ready" : "degraded",
      components,
    });
  });
}
